tensorflow/stream_executor/rocm/rocm_blas.cc

```diff
@@ -1877,37 +1880,70 @@ port::Status ReorganizeMemory(Stream* stream,
   size_t matrix_byte_size = batch_stride * sizeof(MAPPED_T);  
   uint64_t cur_stride_size = matrix_byte_size;
 
+  struct write_op
+  {
+    char* src_ptr;
+    char* dst_ptr;
+    uint64_t size;
+  };
+  std::vector<write_op> write_ops;
+
   for (int i = 1; i < batch_count; ++i) {
     if (reinterpret_cast<char*>(raw_ptrs[i]) == src_ptr + cur_stride_size) {
       cur_stride_size += matrix_byte_size;
     } else {
-      DeviceMemoryBase src_mem = DeviceMemoryBase(src_ptr, cur_stride_size);
-      DeviceMemoryBase target_mem = DeviceMemoryBase(dst_ptr, cur_stride_size);
+      write_ops.push_back(write_op{src_ptr,dst_ptr,cur_stride_size});
+      src_ptr = reinterpret_cast<char *>(raw_ptrs[i]);
+      dst_ptr = device_memory_ptr + i * matrix_byte_size;
+      cur_stride_size = matrix_byte_size;
+    }
+  }  
+  write_ops.push_back(write_op{src_ptr,dst_ptr,cur_stride_size});
+
+  // if the requested operation is reducible to a broadcast (copying the same
+  // buffer multiple times to N uniformly distributed destinations), it may be
```

Comment:
Have you validated it? Or are we able to validate it in some way?

Reply:
If by validated you mean confirmed to work correctly, then yes, the condition is triggered by a number of unit tests, including //tensorflow/python/kernel_tests:einsum_op_test.

Reply:
Oh, I'm actually referring to line 1904 

> it may be more efficient to launch a single broadcast kernel ...

You sounds uncertain about this. So I'm trying to clarify and see if we are able to see deterministic-ally that it brings some benefit.

Reply:
See the link to a spreadsheet I've just added. 

The kernel outperforms the native memcpy loop as long as individual copies are sufficiently small (smaller copies -> framework overhead).

Reply:
Neat! Thanks for your evaluations. This clearly justifies your optimization. We definitely don't want to lose this knowledge. I'd recommend the following:

- Update the comment to describe the data trend (When there are smaller number of copies && each copy has smaller number of bytes, this optimization outperforms).
- Add a confluence page of all excel sheet you have. (No need to be fancy abou the page. Only record the data and briefly talks about the evaluation methodology.) Also refer to the ticket# we file to rocblas. We probably want to re-evaluate rocblas to see if they can outperform your optimization when they have the ticket fixed.

Reply:
To clarify, those benchmarks were only for the code that rearranges the data for its consumption by rocblas, they did not include time spent in rocblas itself. When(if) they properly implement gemm_batched on their end, there will be no need to rearrange the data, so that way would automatically be faster than what we have here.

Reply:
That makes sense, thanks for clarifying.